compiler/src/dotty/tools/dotc/transform/ElimPolyFunction.scala

```diff
@@ -0,0 +1,68 @@
+package dotty.tools.dotc
+package transform
+
+import ast.{Trees, tpd}
+import core._, core.Decorators._
+import MegaPhase._, Phases.Phase
+import Types._, Contexts._, Constants._, Names._, NameOps._, Flags._, DenotTransformers._
+import SymDenotations._, Symbols._, StdNames._, Annotations._, Trees._, Scopes._, Denotations._
+import TypeErasure.ErasedValueType, ValueClasses._
+
+/** This phase rewrite PolyFunction subclasses to FunctionN subclasses
+ *
+ *      class Foo extends PolyFunction {
+ *          def apply(x_1: P_1, ..., x_N: P_N): R = rhs
+ *      }
+ *   becomes:
+ *      class Foo extends FunctionN {
+ *          def apply(x_1: P_1, ..., x_N: P_N): R = rhs
+ *      }
```

Comment:
I wonder if we should allow writing such a class `Foo`, rather than just allowing closures — we don't for implicit function types and nobody seems to mind the restriction, and this restriction enables transformations such as `ShortcutImplicits`. https://github.com/lampepfl/dotty/pull/1775/files#diff-71350811180f41d868e7fb3258fd774dR18

Reply:
Note that it can be really useful being able to extend/implement a polymorphic function type; it's one of the use cases I mention in https://github.com/lampepfl/dotty/issues/4670#issuecomment-397819801 – making polymorphic type classes extend the corresponding polymorphic function type.

Reply:
Ah right, tho `s/type classes/case classes/`. But if we follow the approach for implicit function types, `val b: Int => B[Int] = B` could still work by eta-expansion. In fact, it's not clear why eta-expansion doesn't handle that case today by turning `B` into `B _` or `B.apply` (and I'm not going to try which ones do work, I'd just ask they all work unless backward compatibility gets in the way).

Reply:
@Blaisorblade AFAIK eta expansion never inserts `.apply` calls on things that are not functions. It's true that adding this behavior could be an alternative solution to the stated polymorphic case class problem.